src/daft-core/src/array/mod.rs

```diff
@@ -53,12 +53,30 @@ impl<T> DataArray<T> {
         );
 
         if let Ok(expected_arrow_physical_type) = physical_field.dtype.to_arrow() {
+            // since daft's Utf8 always maps to Arrow's LargeUtf8, we need to handle this special case
+            // If the expected physical type is LargeUtf8, but the actual Arrow type is Utf8, we need to convert it
+            if expected_arrow_physical_type == arrow2::datatypes::DataType::LargeUtf8
+                && arrow_array.data_type() == &arrow2::datatypes::DataType::Utf8
+            {
+                let utf8_arr = arrow_array
+                    .as_any()
+                    .downcast_ref::<arrow2::array::Utf8Array<i32>>()
+                    .unwrap();
+
+                let arr = Box::new(utf8_to_large_utf8(utf8_arr));
+
+                return Ok(Self {
+                    field: physical_field,
+                    data: arr,
+                    marker_: PhantomData,
+                });
+            }
```

Comment:
IMO, we should be able to create a utf8 array from arrow without an explicit cast, but  I can also see the argument for wanting to do this cast outside of the constructor. So I'm fine with if we want to do this outside of `fn new`. 

I think this'll make arrow interop easier as a whole without remembering needing to cast utf8 to largeutf8 every time.

Reply:
cc @andrewgazelka. this should supercede https://github.com/Eventual-Inc/Daft/pull/3602/files#diff-b6317f5746644f4029f8935d6cc782605155701249d1bceaa5e855fe0c13f459R56-R73

Reply:
also @jaychia, @samster25 do you have any preferences on handling this inside the constructor? 

for context, spark uses small utf8, but in rust we don't natively support creating series/DataArray from arrow's smallutf8 array. So this adds a check here to cast the smallutf8 to a largeutf8 when constructing the array. 

See the below test for expected usage.

Reply:
https://github.com/Eventual-Inc/Daft/issues/3605